vyper/ast.py

```diff
@@ -0,0 +1,278 @@
+from itertools import (
+    chain,
+)
+from typing import (
+    Any,
+    List as ListTyping,
+)
+
+from vyper.exceptions import (
+    CompilerPanic,
+)
+
+
+class VyperNode:
+    __slots__ = ('node_id', 'source_code', 'col_offset', 'lineno')
+    ignored_fields = ['ctx', ]
+    only_empty_fields: ListTyping[Any] = []
+
+    @classmethod
+    def get_slots(cls):
+        return chain.from_iterable(
+            getattr(klass, '__slots__', [])
+            for klass in cls.__class__.mro(cls)
+        )
+
+    def __init__(self, **kwargs):
+
+        for field_name, value in kwargs.items():
+            if field_name in self.get_slots():
+                setattr(self, field_name, value)
```

Comment:
I don't think this is actually necessary. 

```python
In [1]: class X:
   ...:     __slots__ = ['foo']
   ...:

In [2]:

In [2]: x = X()

In [3]: x.bar = '3'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-91314b2030ec> in <module>
----> 1 x.bar = '3'

AttributeError: 'X' object has no attribute 'bar'
```

Looks like python enforces this by default.

Reply:
@pipermerriam in this case I want to cacht it very specifically as compiler error, because we as compiler writers' made a mistake if it reaches this point.
That's the whole point of CompilerPanic exception, to indicate to the user that this should not happen.  And is different to an error in their program.